xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/internal/DefaultMacroRefactoring.java

```diff
@@ -0,0 +1,160 @@
+        return Block.LIST_BLOCK_TYPE.equals(contentDescriptor.getType());
+    }
+
+    private Macro<?> getMacro(MacroBlock macroBlock, Syntax syntax) throws MacroRefactoringException
+    {
+        MacroId macroId = new MacroId(macroBlock.getId(), syntax);
+        try {
+            return this.macroManager.getMacro(macroId);
+        } catch (MacroLookupException e) {
+            throw new MacroRefactoringException(String.format("Error while getting macro [%s]", macroId), e);
+        }
+    }
+
+    private Syntax getSyntax(MacroBlock macroBlock, Syntax fallbackSyntax)
+    {
+        Syntax resultSyntax = fallbackSyntax;
+        String metadataSyntax = macroBlock.getParameter(MetaData.SYNTAX);
+
+        if (metadataSyntax != null) {
+            resultSyntax = syntaxRegistry.getSyntax(metadataSyntax).orElse(fallbackSyntax);
+        }
+        return resultSyntax;
+    }
+
+    private XDOM parseMacroContent(MacroBlock macroBlock, Syntax syntax) throws MacroExecutionException
+    {
+        MacroTransformationContext macroTransformationContext = new MacroTransformationContext();
+        macroTransformationContext.setId("refactoring_" + macroBlock.getId());
+        macroTransformationContext.setCurrentMacroBlock(macroBlock);
+        macroTransformationContext.setSyntax(getSyntax(macroBlock, syntax));
+        macroTransformationContext.setInline(macroBlock.isInline());
+
+        return this.macroContentParser
+            .parse(macroBlock.getContent(), macroTransformationContext, false, macroBlock.isInline());
+    }
+
+    private MacroBlock renderMacroBlock(MacroBlock originalBlock, XDOM newXDOMContent, Syntax syntax)
+        throws ComponentLookupException
+    {
+        Syntax contentSyntax = this.getSyntax(originalBlock, syntax);
+        BlockRenderer renderer = this.componentManager.getInstance(BlockRenderer.class, contentSyntax.toIdString());
+        DefaultWikiPrinter printer = new DefaultWikiPrinter();
+        renderer.render(newXDOMContent, printer);
+        String content = printer.toString();
+        return new MacroBlock(originalBlock.getId(), originalBlock.getParameters(), content, originalBlock.isInline());
+    }
+
+    @Override
+    public Optional<MacroBlock> replaceReference(MacroBlock macroBlock, DocumentReference sourceReference,
+        DocumentReference targetReference, DocumentReference currentDocumentReference, Syntax syntax, boolean relative)
+        throws MacroRefactoringException
+    {
+        Macro<?> macro = this.getMacro(macroBlock, syntax);
+        if (this.shouldMacroContentBeParsed(macro)) {
+            try {
+                XDOM xdom = this.parseMacroContent(macroBlock, syntax);
+                List<Block> blocks = this.xdomReferenceRenamer.getBlocks(xdom, false);
+                boolean updated = false;
+                for (Block block : blocks) {
+                    updated |= this.xdomReferenceRenamer.renameReferenceInBlock(block, sourceReference,
```

Comment:
Is it useful to continue the loop once once updated is `true`?

Reply:
It is yes, you need to update all blocks.